connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -181,7 +184,11 @@ public void execute() {
      */
     private boolean sendRecords() {
         int processed = 0;
-        for (final SourceRecord record : toSend) {
+        for (SourceRecord preTransformRecord : toSend) {
+            final SourceRecord record = transformationChain.apply(preTransformRecord);
+
+            if (record == null) continue;
```

Comment:
Interesting -- if you're guaranteed MM-style settings for the producer, then this makes sense. The way I see it there are really two ways you can use `commitRecord`:

1. Assume MM-style settings, i.e. we are basically aiming for the best-effort exactly-once that we can get today w/ infinite retries, 1 request in flight per sec, etc. In this case, *any* acked message indicates everything before it is done and the high watermark thing you're talking about works fine.
2. We might not care about ordering, but have infinite retries (or very large) to try to ensure delivery, i.e. out-of-order at least once semantics. In this case, each `commitRecord` only guarantees you've seen that one record written to Kafka. It is safe to ack exactly that one message (which some input systems will support).

Note that even with the semantics in (1), (2) could still be a relevant due to output shuffling, i.e. if you route by a field, you can't actually use a `commitRecord` as HW for an input partition because two output topics operate effectively independently. In that case, you can only a) ack individual messages on `commitRecord` (which is what, e.g., a traditional MQ might do) or b) you need to track acks on every record and can't use a simple HW tracking. (That second case is basically what the Connect source task is doing for offset commits since we don't know how messages are going to be routed to topics.)

This also raises an interesting point -- it's possible adding transformations after some source connectors will break their prior assumptions about distribution of data across topic partitions in a way that will break their delivery guarantees.